### ISSUE TITLE
Provide docker builds

### ISSUE THREAD

**Issue by DanielFabian** [FStarLang/FStar#584]
I see there are docker builds in the .docker folder, but I can't find the builds on docker hub. Could you set up an automated build?

I am trying to set up a boxed environment with emacs and the settings required for running F* in a docker container and starting from the "official" F* docker would be the obvious choice.

**Comment by beurdouche** [comment]
I don't have much time to look into it now but it is a good idea. 
We could even integrate it in the official repo in the future if it works well...
@DanielFabian Please ping me by the first week of august if I forget.

**Comment by sishtiaq** [comment]
If you're happy to wait until Aug, I can do it too. Assign it to me!

**Comment by beurdouche** [comment]
Sure you can go for it if you want :)

**Comment by s-zanella** [comment]
@DanielFabian: If you have something that works now, we will be happy to integrate it into the repo as the official docker build.
@sishtiaq: if you want to coordinate with Daniel, he's on the 2nd floor!

**Comment by DanielFabian** [comment]
I don't have anything finished just yet. So far, I got it to clone the sources build off the OCaml output, Extract OCaml sources and re-build. 

So essentially quite similar to what the .build folder already has anyway (I only noticed them late). 

No emacs yet, I'm afraid. That said, I'll probably get some time in the evenings / weekend until August, so I'll probably have something by then. 

Once I get at least to usable, I'll make a pull request with what I'll have by then.

**Comment by s-zanella** [comment]
Great! Feel free to discuss it with us, and ask for help if you get stuck.

**Comment by DanielFabian** [comment]
Alright, so here is what I have so far:
[danielfabian/fstar](https://hub.docker.com/r/danielfabian/fstar/)

It's based on my [fstar_devenv](https://github.com/DanielFabian/FStar/tree/add-docker-container/.docker/fstar_devenv) directory. The file is heavily inspired by the [ubuntu_xenial](https://github.com/FStarLang/FStar/tree/master/.docker/ubuntu_xenial) directory but it is not actually `FROM` it. Because the checked in file does not actually build.

What it does:

- Installs OCaml with opam
- Installs mono
- Installs fsharp
- creates a user under which to develop
- clones Z3, rather than taking just a snapshot
- builds and installs Z3 from the correct tag
- clones F*
- adds an `update-fstar.sh` to allow for updating of F* if the image is used in future without re-building it every single time.
- bootstraps F* using F#
- extracts OCaml sources
- builds F* from the extracted sources
- runs the example unit tests
- runs the regression tests

there are a few things that are not great about it so far:

- [ ] The checked in docker files don't work, because they run regeression tests on the *OCaml snapshot* rather than on the freshly extracted OCaml code and the snapshot fails some regression tests. (At least that was true as of a few days ago) 
[line 43](https://github.com/FStarLang/FStar/blob/master/.docker/ubuntu_xenial/Dockerfile#L43) followed by [line 47](https://github.com/FStarLang/FStar/blob/master/.docker/ubuntu_xenial/Dockerfile#L47) without extracting OCaml sources.
- [ ] bootstrapping using OCaml does not work, because the build process uses `fslex` and `fsyacc` which are not provided. Adding the F# dependency. Having F# as a dependency may be fine, if we ever intend to support F# extraction, but it is contradicting the installation manual that claims you can build from the OCaml snapshot.
- [ ] the makefiles assume that under linux .net assemblies are executable, whereas we really need to call them using `mono myassembly.exe`. Leading to a rather nasty `RUNTIME=mono` in the docker file. https://github.com/DanielFabian/FStar/blob/add-docker-container/.docker/fstar_devenv/Dockerfile#L42
- [ ] emacs is not installed and configured
- [ ] vim is not installed and configured
- [ ] atom is not installed and configured
- [ ] should we actually install all 3 editors and set them up? Or possibly provide 3 docker files?
- [ ] it feels like the image is downloading the entire ubuntu repository, twice.
- [ ] if we want to provide images for different distros (or at least different versions of the same distro), we should generate the dockerfile from a template. A simple string replace would probably be just fine; we'd drop the FROM and replace it the specific distro/version we want to run. But this may be more interesting for testing that it builds than for a dev environment?

feedback is welcome.

**Comment by beurdouche** [comment]
Cool ;) Here are a few comments...

Point 1:
This isn't normally supposed to happen as only the OCaml snapshot is required to build F* and this snapshot is always supposed to pass the regression tests.

Regarding regressions, Jonathan has been working on the targets so you can replace the two final lines dedicated to testing with the following one :
`RUN opam config exec -- make -C src regressions` (this is supposed to cover unit tests)

Points 2, 3:
I honestly prefer not to involve anything related to F# as we should not need it anyway but we can certainly have two different dokerfile, one relying on mono and no the other. In that setting points 2 and 3 are irrelevant.

Points 4, 5, 6, 7:
Regarding editors, I think emacs is the highest priority as other editors have F* modes that are not particularly maintained. I would be fine with 3 separate formulas.

Point 8: I have no clue ^^

Point 9: 
Regarding the supported platforms, I now think we should only have a dockerfile for the current Ubuntu LTS. I think it would be too painful to maintain many formulas as I barely managed to maintain 3 of them

WDYT ? B.

**Comment by DanielFabian** [comment]
I like the idea of making the bootstrap work properly from the OCaml sources, it reduces the dependencies, also helping with point 8, as mono is huge. I am more than happy, only providing support for the current ubuntu, as maintaining many different builds is almost certainly going to be painful and probably not too much value in that, especially as this is supposed to be a dev environment.

How about this:
- [ ] adjust the current docker files to properly build from OCaml, not just bootstrap and run unit tests. This will make most of my current script redundant
- [ ] create a dev environment script that's based on the official image. The point being, that in case someone actually feels like installing vim vs. emacs, they have a nice and clean slate and the difference between the official, emacs and vim versions would literally be just the setup of the editor.

**Comment by DanielFabian** [comment]
so if we want to go and try to make the OCaml bootstrap work, how should I deal with the fslex/fsyacc dependency? Just replacing it with ocamllex and ocamlyacc probably won't work, right?

**Comment by beurdouche** [comment]
I am not sure what you mean, we always can build F* in OCaml using the snapshot. Also note that we cannot currently extract to F# if you meant bootstrap that way.

**Comment by DanielFabian** [comment]
building F*, as per `make -C src/ocaml-output` works just fine, when OCaml is around. What doesn't work, is extracting to OCaml `make ocaml -C src`.

Unless I misunderstood what those two targets mean.

The way how I understood it, I should be able to:
- `make -C src/ocaml-output` (builds an F* compiler from the slightly outdated OCaml snapshot)
- `make ocaml -C src` (extracts OCaml sources from the most current F* compiler - this fails without fslex/fsyacc)
- `make -C src/ocaml-output` (builds the F* compiler from the freshly extracted OCaml sources)

**Comment by beurdouche** [comment]
Yes, this is indeed what happens. The thing is that before that we always do `make -C src` that requires mono to get FsLexYacc from Nuget. This means that we cannot update the OCaml snapshot before building F*. That said, I believe it is acceptable to build from a (potentially) slightly outdated ML source code. It is usually the case that people changing the compiler will update the ML snapshot accordingly. Only in rare occasions the build would fail...
```
make -C src/
mono VS/.nuget/NuGet.exe restore VS/FStar.sln
Installing 'FsLexYacc 6.1.0'.
Installing 'FsLexYacc.Runtime 6.1.0'.
Successfully installed 'FsLexYacc.Runtime 6.1.0'.
Successfully installed 'FsLexYacc 6.1.0'.
```

**Comment by DanielFabian** [comment]
Doesn't that actually mean, the OCaml snapshot can't be updated without mono? Because the `make ocaml -C src` requires mono... I think I am missing something.

**Comment by DanielFabian** [comment]
Alright, I looked at it more closely and here is what's wrong:

https://github.com/FStarLang/FStar/blob/master/src/Makefile#L132 This line expects a `parser.fsi` but that's not around, as it get generated during the F# build process using fsyacc. I think, we would need to modify the Makefile, such that it runs ocamlyacc instead for the OCaml extraction.

**Comment by s-zanella** [comment]
It might not be as simple as running ocamlyacc because you would also need to translate the resulting .ml/.mli.

**Comment by msprotz** [comment]
Had a chat with @s-zanella and @DanielFabian:
- `parse.fsi` is now under version control to support ocaml-snapshot compilation, re-extraction, re-compilation without mono
- ocaml_extraction in Makefile is wrong: it always, unconditionally, prepends $RUNTIME.

The fix is as follows:
- remove RUNTIME from the Makefile
- check in bin/fstar-any.sh with the following contents:

```
#!/bin/bash
FSTAR=$(which fstar.exe)
if (( $? != 0 )); then
  echo "fstar.exe not found"
elif file $FSTAR | grep Mono >/dev/null 2>&1; then
 mono --debug $FSTAR "$@"
else
 $FSTAR "$@"
fi
```

then, in the Makefile (pseudo-code):

```
ifeq($(UNAME),Windows_NT)
FSTAR=../bin/fstar.exe
else
FSTAR=../bin/fstar-any.sh
```

then, replace every invocation of `../bin/fstar.exe` in `src/Makefile` with `$(FSTAR)`

Daniel will implement the latter bit. I've implemented the checkin of parse.fsi

Thanks Daniel!!

**Comment by DanielFabian** [comment]
Based on the Makefile changes, I now have a fresh docker file, much in the spirit of the currently check-in ones. With a few improvements targeted at interactive development.

- It creates a user account
- clones z3
- bootstraps, extracts and re-builds in pure OCaml
- adds a tiny script to pull from the upstream repo and re-build F*.

Once the Makefile changes are merged, I can make a pull request for this and it can serve as a baseline before adding editors.

**Comment by DanielFabian** [comment]
With the PR in, I think, the last thing to do is create a dockerhub account and set up the CI for it. Which is something, the F* team needs to do.

**Comment by markulf** [comment]
@vkanne-msft what is the status of this compared to the `mitls-fstar` docker setup?

**Comment by markulf** [comment]
How difficult would it be to create an image that is specifically targeting tutorial users that want to take it to the next level by running F* on the tutorial examples outside the web interface?

**Comment by beurdouche** [comment]
@DanielFabian we have an organisation here: https://hub.docker.com/r/fstarlang/fstar/
Can you send me your login on the Everest Slack so that I can give you permissions ?

**Comment by DanielFabian** [comment]
@markulf What features do you think are missing for tutorial users? With https://github.com/FStarLang/FStar/pull/592 you get a fully set up F* docker with emacs pre-installed and pre-configured. Furthermore the whole F* repo is cloned, so in particular, the F* samples are there? Also when we do upload them to dockerhub, CI will run on it, so the tutorial won't ever be out-of-date.

**Comment by DanielFabian** [comment]
@beurdouche I'll contact you on Tuesday when I get back to the office.

**Comment by markulf** [comment]
@DanielFabian this looks like exactly what is needed. Do you have installation instructions somewhere? Maybe add them to the tutorial.

**Comment by DanielFabian** [comment]
@markulf I'll look into it next week.

**Comment by sishtiaq** [comment]
@markulf @beurdouche This issue can be closed, right?

**Comment by beurdouche** [comment]
@sishtiaq, I think we miss the vim and atom modes but honestly I am not sure it is necessary. @DanielFabian what do you think about that ? closing ?

**Comment by beurdouche** [comment]
Also @msprotz, I know you use vim, WDYT ?

**Comment by msprotz** [comment]
no opinion... I don't even use the fstar mode...

**Comment by A-Manning** [comment]
@beurdouche the Atom mode is still maintained, please check the PRs.

**Comment by sishtiaq** [comment]
This issue is to track having a Docker image for doing what we do everyday on Linux or Windows+Cygwin. I think the images that Danny, Benjamin and Markulf worked on a few weeks ago do this. So, I think we can close this issue.

**Comment by beurdouche** [comment]
@A-Manning Sure, but as @sishtiaq said it is more a reference environnement than something having support for all editor configurations. That said, if someone is willing to maintain an F* + atom Dockerfile, why not, this is a slightly different problem. I am very interested in waiting on @DanielFabian opinion here. but we can do that in another issue...

**Comment by beurdouche** [comment]
Closing in favor of https://github.com/FStarLang/FStar/issues/716

**Comment by sishtiaq** [comment]
Thanks @Beurdouche!

**Comment by beurdouche** [comment]
(For reference)
